src/core/symbology-ng/qgssymbolv2.cpp

```diff
@@ -500,13 +658,207 @@ bool QgsSymbolV2::hasDataDefinedProperties() const
+        static_cast<QgsMarkerSymbolV2*>( this )->renderPoint( pt, &feature, context, layer, selected );
+      }
+    }
+    break;
+
+    case QgsWKBTypes::MultiCurve:
+    case QgsWKBTypes::MultiLineString:
+    {
+      QPolygonF pts;
+
+      if ( mType != QgsSymbolV2::Line )
+      {
+        QgsDebugMsg( "multi-linestring can be drawn only with line symbol!" );
+        break;
+      }
+
+      QgsConstWkbPtr wkbPtr( segmentizedGeometry->asWkb() + 1 + sizeof( int ) );
+      unsigned int num;
+      wkbPtr >> num;
+      const unsigned char* ptr = wkbPtr;
+
+      const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );
+
+      for ( unsigned int i = 0; i < num; ++i )
+      {
+        if ( geomCollection )
+        {
+          context.setGeometry( geomCollection->geometryN( i ) );
+        }
+        ptr = QgsConstWkbPtr( _getLineString( pts, context, ptr, clipFeaturesToExtent() ) );
+        static_cast<QgsLineSymbolV2*>( this )->renderPolyline( pts, &feature, context, layer, selected );
+      }
+    }
+    break;
+
+    case QgsWKBTypes::MultiSurface:
+    case QgsWKBTypes::MultiPolygon:
+    {
+      if ( mType != QgsSymbolV2::Fill )
+      {
+        QgsDebugMsg( "multi-polygon can be drawn only with fill symbol!" );
+        break;
+      }
+
+      QgsConstWkbPtr wkbPtr( segmentizedGeometry->asWkb() + 1 + sizeof( int ) );
+      unsigned int num;
+      wkbPtr >> num;
+      const unsigned char* ptr = wkbPtr;
+
+      QPolygonF pts;
+      QList<QPolygonF> holes;
+
+      const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );
+
+      for ( unsigned int i = 0; i < num; ++i )
+      {
+        if ( geomCollection )
+        {
+          context.setGeometry( geomCollection->geometryN( i ) );
+        }
+        ptr = _getPolygon( pts, holes, context, ptr, clipFeaturesToExtent() );
+        static_cast<QgsFillSymbolV2*>( this )->renderPolygon( pts, ( holes.count() ? &holes : NULL ), &feature, context, layer, selected );
+      }
+      break;
+    }
+    default:
+      QgsDebugMsg( QString( "feature %1: unsupported wkb type 0x%2 for rendering" ).arg( feature.id() ).arg( geom->wkbType(), 0, 16 ) );
+  }
+
+  if ( drawVertexMarker )
+  {
+    const QgsCoordinateTransform* ct = context.coordinateTransform();
+    const QgsMapToPixel& mtp = context.mapToPixel();
+
+    QgsPointV2 vertexPoint;
+    QgsVertexId vertexId;
+    double x, y, z;
+    QPointF mapPoint;
+    while ( geom->geometry()->nextVertex( vertexId, vertexPoint ) )
+    {
+      //transform
+      x = vertexPoint.x(); y = vertexPoint.y(); z = vertexPoint.z();
+      if ( ct )
+      {
+        ct->transformInPlace( x, y, z );
+      }
+      mapPoint.setX( x ); mapPoint.setY( y );
+      mtp.transformInPlace( mapPoint.rx(), mapPoint.ry() );
+      QgsVectorLayer::drawVertexMarker( mapPoint.x(), mapPoint.y(), *context.painter(),
```

Comment:
hmm... interesting. It looks to me like this is drawing every vertex, even if it's way outside of the visible bounds. We should test if it's inside the canvas extent and avoid drawing the marker otherwise. Could potentially be a big performance boost!